templates/cc/register.go

```diff
@@ -268,14 +272,53 @@ func cType(t pgs.FieldType) string {
 	if t.IsEmbed() {
 		return className(t.Embed())
 	}
-	if t.IsRepeated() {
-		// Strip the leading []*
-		return cTypeOfString(t.Name().String())[3:]
+	if t.IsRepeated(){
+		if t.ProtoType() == pgs.MessageT {
+			return className(t.Element().Embed())
+		}
+		// Strip the leading []
+		return cTypeOfString(t.Name().String()[2:])
 	}
 
 	return cTypeOfString(t.Name().String())
 }
 
+// Compute unique C++ types that correspond to all message fields in a
+// compilation unit that need to be weak (i.e. not already defined). Used to
+// generate weak default definitions for CheckMessage.
+func weakCheckMsgs(msgs []pgs.Message) []string {
+	already_defined := map[string]bool{}
+	// First compute the C++ type names for things we're going to provide an explicit
+	// CheckMessage() with Validate(..) body in this file. We can't define the
+	// same CheckMessage() signature twice in a compilation unit, even if one of
+	// them is weak.
+	for _, msg := range msgs {
+		already_defined[className(msg)] = true
+	}
+	// Compute the set of C++ type names we need weak definitions for.
+	ctype_map := map[string]bool{}
+	for _, msg := range msgs {
+		if disabled, _ := shared.Disabled(msg); disabled {
+			continue
+		}
+		for _, f := range msg.Fields() {
+			ctype := cType(f.Type())
+			if already_defined[ctype] {
+				continue
+			}
+			if f.Type().IsEmbed() || (f.Type().IsRepeated() && f.Type().Element().IsEmbed()) {
+				ctype_map[ctype] = true
+			}
+		}
+	}
+	// Convert to array.
+	ctypes := []string{}
```

Comment:
nit: preallocate a slice with a large enough capacity using `make()` since the size is known